### PR TITLE
A tutorial for the ingress

### LINKED ISSUE #31: Ingress Integration Docs
As part of https://github.com/fission/fission/issues/164 we need docs/updates on how ingress will work

### PR DESCRIPTION
**Pull request by vishal-biyani**
Fixes #31

### CODE REVIEW

File: docs/content/tutorial/ingress-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
+---
+title: "Exposing functions with Ingress"
+draft: false
+weight: 43
+---
+
+This tutorial will walk you through exposing a function using a ingress controller. We will make the function available on a FQDN using Fission's route and ingress controller setup in a cloud environment.
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: Okay maybe I am wrong about this but, I think it should be "an ingress"

Comment (erwinvaneyk) [reply docs/content/tutorial/ingress-tutorial.md]: Maybe add a link to a generic explanation of what an ingress controller is. Maybe some page in the kubernetes guide?

File: docs/content/tutorial/ingress-tutorial.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+## Setup & pre-requisites
+
+You will need a Kubernetes cluster with Fission installed (Please check [installation page](../../installation/installation/) for details). This tutorial uses a cloud load balancer and if you are trying on Minikube, you might want [to take a look at details here](https://github.com/kubernetes/minikube/issues/496)
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: balancer and if you ->  balancer, but if you are using Minikube you might want

File: docs/content/tutorial/ingress-tutorial.md (hunk 1 above)
Comment (erwinvaneyk) [reply docs/content/tutorial/ingress-tutorial.md]: Unroll the FQDN acronym here. I needed to look it up what it exactly meant.

File: docs/content/tutorial/ingress-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+Later parts of this tutorial use a FQDN to reach the function. If you plan to go along in this section, you will need a domain name setup and access to modify the NS records & create A record in the zone of the domain name you have. The tutorial uses Google cloud to walk through the tutorial but you can use any cloud you prefer to. Also the changes in name server can take upto 24-48 hours so you may want to use an already created domain name.
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: & -> and

Comment (erwinvaneyk) [reply docs/content/tutorial/ingress-tutorial.md]: remove upto

File: docs/content/tutorial/ingress-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+### Setup an Ingress Controller
+
+First thing we will need is an ingress controller and we will use Nginx ingress controller in this tutorial. Based on your setup you can choose [one of the ways to install Nginx ingress controller from here](https://kubernetes.github.io/ingress-nginx/deploy/). This setup should work with other ingress controllers also but has not been tested.
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: one of the ways to install Nginx ingress controller from here -> one of multiple ways to install Nginx ingress controller 

Try to avoid using "click here" or "from here" in links.

File: docs/content/tutorial/ingress-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+Once ingress controller is installed properly, let's verify that the installation succeeded (In my case all components of ingress controller were installed in ingress-nginx namespace):
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: This sentence sounds like a tautology. Maybe rephrase?

File: docs/content/tutorial/ingress-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+NAME                              DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+deploy/default-http-backend       1         1         1            1           19d
+deploy/nginx-ingress-controller   1         1         1            1           19d
+
+NAME                                     DESIRED   CURRENT   READY     AGE
+rs/default-http-backend-66b447d9cf       1         1         1         19d
+rs/nginx-ingress-controller-58fcfdc6fd   1         1         1         19d
+
+```
+Some key points to validate that ingress controller installation succeeded:
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: points -> pointers

Comment (erwinvaneyk) [reply docs/content/tutorial/ingress-tutorial.md]: Remove 'Some'

File: docs/content/tutorial/ingress-tutorial.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+- The ingress controller pod is up and running
+- The ingress-nginx service has a external IP address populated
+- If you hit the external IP address of the ingress-nginx, you get the default backend page:
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: You could add some example commands for each of the pointers (in parentheses). For the first two ones.

File: docs/content/tutorial/ingress-tutorial.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+```
+curl http://35.200.150.175
+
+default backend - 404
+```
+
+## Deploying Function with ingress
+
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: Add some text here about the purpose of this section

File: docs/content/tutorial/ingress-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+### Create a function
+
+We will create an environment, function and test that it works:
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: This misses context (see previous comment).

File: docs/content/tutorial/ingress-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+function 'hello' created
+
+$ fission fn test --name hello
+Hello, Fission!
+
+```
+
+### Create a internal route
+
+Let's create a route which is not exposed via the ingress controller:
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: Same as previous two comments, explain why we need to do this. What will happen.

File: docs/content/tutorial/ingress-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+```
+If you check the ingress controller pod logs, you will notice that the ingress controller has re-loaded the configuration for the newly created ingress resource:
+
+```
+I0604 12:47:08.983567       5 controller.go:168] backend reload required
+I0604 12:47:08.985535       5 event.go:218] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"fission", Name:"301b3cb0-5ac1-4211-a1ed-2b0ad9143e34", UID:"64bffe8c-67f5-11e8-98e8-42010aa00018", APIVersion:"extensions", ResourceVersion:"18017617", FieldPath:""}): type: 'Normal' reason: 'CREATE' Ingress fission/301b3cb0-5ac1-4211-a1ed-2b0ad9143e34
+I0604 12:47:09.117629       5 controller.go:177] ingress backend successfully reloaded...
+```
+
+If you now hit the function at ingress controller's IP and the path (`http://<INGRESS-CONTROLLER-EXTERNAL-IP>/hello`), you will get function response. Depending on your setup and settings, you will have to try HTTP or HTTPS. Some ingress controllers by default enable SSL redirect and hence HTTPS URL has to be accessed.
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: you will get function response -> you will get the function's response

Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: Some ingress controllers by default enable SSL redirect and hence HTTPS URL has to be accessed. -> Some ingress controllers enable SSL redirect by default and hence the HTTPS URL has to be accessed.

File: docs/content/tutorial/ingress-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+```
+$ curl -k  https://35.200.150.175/hello
+Hello, Fission!
+
+```
+
+### Create a FQDN route (Optional, needs pre-requisites fulfilled)
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: leave out 'needs pre-requisites fulfilled'. Address this in the text

File: docs/content/tutorial/ingress-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+Lastly we can map the FQDN to function if you have DNS setup and access. You need to do a few steps:
+
+- Map the domain name's name server to your cloud provider. For example we used domain name fission.sh and mapped the name server to google cloud (Since this tutorial setup is on Google cloud)
```
Comment (erwinvaneyk) [inline docs/content/tutorial/ingress-tutorial.md]: Show an example of the correct records here for the example (fission.sh).

File: docs/content/tutorial/ingress-tutorial.md (hunk 7 above)
Comment (vishal-biyani) [reply docs/content/tutorial/ingress-tutorial.md]: Since commands & names might vary, will point to output from previous command instead of showing commands here.

File: docs/content/tutorial/ingress-tutorial.md (hunk 13 above)
Comment (vishal-biyani) [reply docs/content/tutorial/ingress-tutorial.md]: This gets very specific to the domain name provider, will mention that one needs to check the docs of domain name provider.
